### PR TITLE
[MBL-1453] PagedTabBar implementation and adding to PagedContainerViewController

### PR DESCRIPTION
**Pull request by stevestreza-ksr**
# 📲 What

image

Reimplements the paged tab bar look in SwiftUI, integrates it into Amy's existing PagedContainerViewController, and adds tests for it.

# 🛠 How

The `PagedTabBar` takes a list of instances of `TabBarPage` items, a basic protocol that has a name and badge count. It uses the `PagedContainerViewModel` for determining which tab to show.

The `PagedContainerViewModel` was cleaned up and had nearly all calls to index-based pages removed in favor of returning the TabBarPage instances. This should prevent worrying about over-indexing and risking runtime crashes.

The new `PagedTabBar` has some screenshot tests and the `PagedContainerViewModel` has had some unit tests added.

Note: Originally there was a separate PagedTabBarViewModel that worked together with the PagedContainerViewModel, but this ended up being way too easy to accidentally end up with an infinitely recursive scenario where the PagedTabBarViewModel called the PagedContainerViewModel which called the PagedTabBarViewModel which called … [etc]. Since they are both relying on reactivity to update UI, for now they just use the same view model.

# 👀 See

[Figma](https://www.figma.com/design/r0WKSECmMTCJTSKQskt2SQ/Backed-Projects-Dashboard?node-id=722-8194&t=sNPDyrLmcinFhNT1-0) and the screenshot at the top of the screen

# ♿️ Accessibility 

- [x] Tap targets use minimum of 44x44 pts dimensions
- [x] Works with VoiceOver
- [x] Supports Dynamic Type 

# ✅ Acceptance criteria

- [x]Built in SwiftUI or UIKit (developer’s choice; the container screen is a UIViewController)
- **N/A**: Could possibly be reused from SortPagerViewController
- [x] Gracefully displays arbitrary titles and badges, even with long tab names (i.e. German)
- [x] Supports basic accessibility needs (adaptive fonts, VoiceOver)
- [ ] Animates pleasingly when tabs are switched 
- [x] Provides an appropriate delegate / callback / binding method so that a view using the tab bar knows when the tabs are swapped

# ⏰ TODO

- [ ] Animation. Tried a few techniques for doing this but was having trouble getting SwiftUI to not break when doing them. May have to spin this into a fast-follow task.

### CODE REVIEW

**Review by ifosli — CHANGES_REQUESTED** [review]
Nice! I left a few comments, but overall I like it! I am quite concerned by the reliance on a badge number, though, since we're not going to get that on the user object for milestone 1. It might be worth confirming with Alison what we want this to look like for now before getting rid of it entirely, though. (I'm not actually sure if we'll want to show the count on the activity tab only - it'd be useful info but it'd look kinda weird.) If you do have an idea for how to get the badge number for the project alerts I'm happy to approve this pr as is; just let me know!

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

### SwiftFormat found issues:

| File | Rules |
| ---- | ----- |
| Library/PagedContainer/PagedContainerViewController.swift:33:1 | warning: (trailingClosures) Use trailing closure syntax where applicable. |
| Library/PagedContainer/PagedContainerViewModel.swift:26:1 | warning: (wrap) Wrap lines that exceed the specified maximum width. |
| Library/PagedContainer/PagedContainerViewModel.swift:42:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedContainerViewModel.swift:48:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:15:1 | warning: (wrap) Wrap lines that exceed the specified maximum width. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:17:1 | warning: (trailingClosures) Use trailing closure syntax where applicable. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:19:1 | warning: (trailingClosures) Use trailing closure syntax where applicable. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:24:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:28:1 | warning: (trailingClosures) Use trailing closure syntax where applicable. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:30:1 | warning: (redundantParens) Remove redundant parentheses. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:34:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:46:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:63:1 | warning: (trailingClosures) Use trailing closure syntax where applicable. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:65:1 | warning: (redundantParens) Remove redundant parentheses. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:69:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:77:1 | warning: (trailingClosures) Use trailing closure syntax where applicable. |
| Library/PagedContainer/PagedContainerViewModelTests.swift:84:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedTabBar.swift:9:1 | warning: (redundantParens) Remove redundant parentheses. |
| Library/PagedContainer/PagedTabBarTests.swift:27:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedTabBarTests.swift:47:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedTabBarTests.swift:63:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedTabBarTests.swift:79:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |
| Library/PagedContainer/PagedTabBarTests.swift:95:1 | warning: (redundantSelf) Insert/remove explicit self where applicable. |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Kickstarter-iOS/Features/PledgedProjectsOverview/PPOContainerViewController.swift [hunk 1]
```diff
@@ -16,6 +16,37 @@ public class PPOContainerViewController: PagedContainerViewController {
     let activitiesViewController = ActivitiesViewController.instantiate()
     activitiesViewController.title = "Activity Feed"
 
-    self.setPagedViewControllers([ppoViewController, activitiesViewController])
+    self.setPagedViewControllers([
+      (.projectAlerts(badge: 5), ppoViewController),
```
[the remark this answers is not in the document]
Comment (stevestreza-ksr) [reply Kickstarter-iOS/Features/PledgedProjectsOverview/PPOContainerViewController.swift]: Added a basic dot style alongside the badge count style, along with an enum for the different badge styles.

File: Library/PagedContainer/PagedTabBar.swift [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  private func isSelected(page: Page) -> Bool {
+    self.selection?.id == page.id
+  }
+}
+
+private enum Constants {
+  static let height: CGFloat = 46
+  static let lineLimit = 1
```
[the remark this answers is not in the document]
Comment (stevestreza-ksr) [reply Library/PagedContainer/PagedTabBar.swift]: Tests for big text labels added. Two lines fits under normal text circumstances so I added that as well, and height constraints will take care of it at larger font sizes.

**Pull request merged**